### ISSUE TITLE
Isabelle Recursive Functions

### ISSUE THREAD

**Issue by Barrometer** [johnwickerson/HSV#5]
With task 10 and defining constant folding I have run into an issue, which I think is due how to Isabelle handles recursive functions.

I want to write a line similar to 
`| "opt_folding (NOT c) = opt_folding (NOT opt_folding c)" `
This line is to handle the general case for folding around a NOT. However, if I include this line, Isabelle returns a large error, of which I think the relevant line is "Could not find lexicographic termination order."

I originally tried the line 
`| "constant_folding (NOT c) = NOT (constant_folding c)"`
However this doesn't properly handle cases such as 
`value "constant_folding(NOT(AND TRUE TRUE))"` which evaluates as "NOT TRUE".

Is there a way to change how Isabelle attempts to parse and prove recursive functions so as to complete Task 10?

**Comment by johnwickerson** [comment]
Great question! 

When you declare a recursive function, Isabelle immediately proves that every call to that function will always terminate. This usually happens invisibly because termination proofs are usually easy.

But in this case, termination is not obvious. The argument given to your function in the recursive call is not obviously “smaller” than the original argument. 

There are ways to supply Isabelle with more complicated termination proofs. You replace “fun” with “function” and then you can write “termination proof ....” after the definition. 

But I would first think about whether your function *actually* terminates on all inputs, and if you’re not sure that it does, consider rewriting it.

**Comment by Barrometer** [comment]
I am reasonably sure that for a finite sized input circuit, it should eventually terminate given my other rules / aspects of the function definition.

Another approach I tried is to write a second function, which wraps around opt_folding, and applies opt_folding successively until it is either TRUE, FALSE, INPUT i, or NOT(INPUT i) and could not get this to work either. Is such a wrapper a more sensible way to handle this?

Also, is there a way to change how Isabelle applies a function to an input? If the function were to be applied to the "roots" of the circuit first, and then propagate upwards, I think this function would be easier to write.

**Comment by johnwickerson** [comment]
> I am reasonably sure that for a finite sized input circuit, it should eventually terminate given my other rules / aspects of the function definition.

Yes you can always assume that the input circuit has a finite size. The definition of `circuit` does not provide for the possibility of writing down an infinitely big circuit.

> Another approach I tried is to write a second function, which wraps around opt_folding, and applies opt_folding successively until it is either TRUE, FALSE, INPUT i, or NOT(INPUT i) and could not get this to work either. Is such a wrapper a more sensible way to handle this?

Yes, I reckon so.

> Also, is there a way to change how Isabelle applies a function to an input? If the function were to be applied to the "roots" of the circuit first, and then propagate upwards, I think this function would be easier to write.

You can control the order in which the circuit is traversed by your placement of the recursive calls. If you have a recursive function that looks something like this:

```isabelle
fun f where
"f (Not c) = blah (f (foo c))"
| ...
```

then you can control whether the function works "root first" or "leaves first" by choosing whether you do the bulk of the work *before* the recursive call (i.e., in `foo`) or *after* the recursive call (i.e., in `blah`).
